web_poet/rules.py

```diff
@@ -115,9 +116,63 @@ class ExampleComProductPage(WebPage[Product]):
     """
 
     def __init__(self, *, rules: Optional[Iterable[ApplyRule]] = None):
-        self._rules: List[ApplyRule] = []
+        self._rules: Dict[int, ApplyRule] = {}
+        self._overrides_matchers: Dict[
+            Optional[Type[ItemPage]], URLMatcher
+        ] = defaultdict(URLMatcher)
+        self._item_matchers: Dict[Optional[Type], URLMatcher] = defaultdict(URLMatcher)
```

Comment:
Why is Optional used here, are there cases when keys are going to be None?
Also, if it's a defaultdict, should't we use defaultdict as a type, not Dict?

Reply:
In https://github.com/scrapinghub/web-poet/pull/112/commits/dc6181024758aa0c1038a70b03a6f529f479c327, it allows faster look-up on cases like `.search(to_return=None)`. Combos well with `instead_of` like `.search(to_return=None, instead_of=None)`.

Updated to use `DefaultDict`.